framework/src/modules/chain/utils/sequence.js

```diff
@@ -28,81 +35,76 @@ const util = require('util');
  * @todo Add description for the params
  */
 class Sequence {
-	constructor(config) {
-		let _default = {
-			onWarning: null,
-			warningLimit: 50,
+	constructor({ onWarning, warningLimit } = defaultConfig) {
+		this.queue = [];
+		this.config = {
+			onWarning,
+			warningLimit,
 		};
-		_default = Object.assign(_default, config);
-		const self = this;
-		this.sequence = [];
 
-		setImmediate(function nextSequenceTick() {
-			if (_default.onWarning && self.sequence.length >= _default.warningLimit) {
-				_default.onWarning(self.sequence.length, _default.warningLimit);
+		const nextSequence = async () => {
+			if (
+				this.config.onWarning &&
+				this.queue.length >= this.config.warningLimit
+			) {
+				this.config.onWarning(this.queue.length, this.config.warningLimit);
 			}
-			self.__tick(() => {
-				setTimeout(nextSequenceTick, 3);
-			});
-		});
+			await this._tick();
+			setTimeout(nextSequence, defaultTickInterval);
+		};
+
+		nextSequence();
 	}
 
 	/**
-	 * Removes the first task from sequence and execute it with args.
+	 * Removes the first task from queue and execute it with args.
 	 *
 	 * @param {function} cb
 	 * @returns {setImmediateCallback} With cb or task.done
 	 * @todo Add description for the params
 	 */
-	__tick(cb) {
-		const task = this.sequence.shift();
+	async _tick() {
+		const task = this.queue.shift();
 		if (!task) {
-			return setImmediate(cb);
+			return;
 		}
-		let args = [
-			function(err, res) {
-				if (task.done) {
-					setImmediate(task.done, err, res);
-				}
-				setImmediate(cb);
-			},
-		];
-		if (task.args) {
-			args = args.concat(task.args);
+		try {
+			const result = await task.worker();
+			task.done.resolve(result);
+		} catch (error) {
+			task.done.reject(error);
 		}
-		return task.worker.apply(task.worker, args);
 	}
 
 	/**
-	 * Adds a new task to sequence.
+	 * Adds a new task to queue.
 	 *
 	 * @param {function} worker
 	 * @param {Array} args
 	 * @param {function} done
 	 * @todo Add description for the params
 	 */
-	add(worker, args, done) {
-		if (!done && args && typeof args === 'function') {
-			done = args;
-			args = undefined;
-		}
-		if (worker && typeof worker === 'function') {
-			const task = { worker, done };
-			if (util.isArray(args)) {
-				task.args = args;
-			}
-			this.sequence.push(task);
+	add(worker) {
+		if (!util.types.isAsyncFunction(worker)) {
+			throw new Error('Worker must be an async function.');
 		}
+		let done;
+		const workerPromise = new Promise((resolve, reject) => {
+			done = { resolve, reject };
+		});
+		const task = { worker, done };
+		this.queue.push(task);
```

Comment:
We can simply wrap the promise and remove the need to use closures. Like,
```
const workerPromise = new Promise((resolve, reject) => {
 try {
  resolve(await worker);
 } catch (e) {
  reject(e);
 }
});
this.queue.push(workerPromise);
return workerPromise;
```

Reply:
When the tick is called, we want to make sure that it doesn't throw, but it resolve/reject the workerPromise.
If we do this way, I wasn't sure how to do that